File: tests/rosetta/transpiler/php/100-prisoners.php
<?php
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
$__start_mem = memory_get_usage();
$__start = _now();
  function mochi_shuffle($xs) {
  $arr = $xs;
  $i = 99;
  while ($i > 0) {
  $j = fmod(_now(), ($i + 1));
  $tmp = $arr[$i];
  $arr[$i] = $arr[$j];
  $arr[$j] = $tmp;
  $i = $i - 1;
};
  return $arr;
};
  function doTrials($trials, $np, $strategy) {
  $pardoned = 0;
  $t = 0;
  while ($t < $trials) {
  $drawers = [];
  $i = 0;
  while ($i < 100) {
  $drawers = array_merge($drawers, [$i]);
  $i = $i + 1;
};
  $drawers = mochi_shuffle($drawers);
  $p = 0;
  $success = true;
  while ($p < $np) {
  $found = false;
  if ($strategy == 'optimal') {
  $prev = $p;
  $d = 0;
  while ($d < 50) {
  $mochi_this = $drawers[$prev];
  if ($mochi_this == $p) {
  $found = true;
  break;
}
  $prev = $mochi_this;
  $d = $d + 1;
};
} else {
  $opened = [];
  $k = 0;
  while ($k < 100) {
  $opened = array_merge($opened, [false]);
  $k = $k + 1;
};
  $d = 0;
  while ($d < 50) {
  $n = fmod(_now(), 100);
  while ($opened[$n]) {
  $n = fmod(_now(), 100);
};
  $opened[$n] = true;
  if ($drawers[$n] == $p) {
  $found = true;
  break;
}
  $d = $d + 1;
};
}
  if (!$found) {
  $success = false;
  break;
}
  $p = $p + 1;
};
  if ($success) {
  $pardoned = $pardoned + 1;
}
  $t = $t + 1;
};
  $rf = (floatval($pardoned)) / (floatval($trials)) * 100.0;
  echo rtrim('  strategy = ' . $strategy . '  pardoned = ' . _str($pardoned) . ' relative frequency = ' . _str($rf) . '%'), PHP_EOL;
};
  function main() {
  $trials = 1000;
  foreach ([10, 100] as $np) {
  echo rtrim('Results from ' . _str($trials) . ' trials with ' . _str($np) . ' prisoners:
'), PHP_EOL;
  foreach (['random', 'optimal'] as $strat) {
  doTrials($trials, $np, $strat);
};
};
};
  main();
$__end = _now();
$__end_mem = memory_get_usage();
$__duration = intdiv($__end - $__start, 1000);
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;;

File: tests/rosetta/transpiler/Clojure/100-prisoners.clj
(ns main (:refer-clojure :exclude [shuffle doTrials main]))

(require 'clojure.set)

(def nowSeed (atom (let [s (System/getenv "MOCHI_NOW_SEED")] (if (and s (not (= s ""))) (Integer/parseInt s) 0))))

(declare shuffle doTrials main)

(defn shuffle [xs]
  (try (do (def arr xs) (def i 99) (while (> i 0) (do (def j (mod (swap! nowSeed (fn [s] (mod (+ (* s 1664525) 1013904223) 2147483647))) (+ i 1))) (def tmp (nth arr i)) (def arr (assoc arr i (nth arr j))) (def arr (assoc arr j tmp)) (def i (- i 1)))) (throw (ex-info "return" {:v arr}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn doTrials [trials np strategy]
  (do (def pardoned 0) (def t 0) (while (< t trials) (do (def drawers []) (def i 0) (while (< i 100) (do (def drawers (conj drawers i)) (def i (+ i 1)))) (def drawers (shuffle drawers)) (def p 0) (def success true) (loop [while_flag_1 true] (when (and while_flag_1 (< p np)) (do (def found false) (if (= strategy "optimal") (do (def prev p) (def d 0) (loop [while_flag_2 true] (when (and while_flag_2 (< d 50)) (do (def this (nth drawers prev)) (cond (= this p) (do (def found true) (recur false)) :else (do (def prev this) (def d (+ d 1)) (recur while_flag_2))))))) (do (def opened []) (def k 0) (while (< k 100) (do (def opened (conj opened false)) (def k (+ k 1)))) (def d 0) (loop [while_flag_3 true] (when (and while_flag_3 (< d 50)) (do (def n (mod (swap! nowSeed (fn [s] (mod (+ (* s 1664525) 1013904223) 2147483647))) 100)) (while (nth opened n) (def n (mod (swap! nowSeed (fn [s] (mod (+ (* s 1664525) 1013904223) 2147483647))) 100))) (def opened (assoc opened n true)) (cond (= (nth drawers n) p) (do (def found true) (recur false)) :else (do (def d (+ d 1)) (recur while_flag_3)))))))) (cond (not found) (do (def success false) (recur false)) :else (do (def p (+ p 1)) (recur while_flag_1)))))) (when success (def pardoned (+ pardoned 1))) (def t (+ t 1)))) (def rf (* (/ (double pardoned) (double trials)) 100)) (println (str (str (str (str (str (str "  strategy = " strategy) "  pardoned = ") (str pardoned)) " relative frequency = ") (str rf)) "%"))))

(defn main []
  (do (def trials 1000) (doseq [np [10 100]] (do (println (str (str (str (str "Results from " (str trials)) " trials with ") (str np)) " prisoners:\n")) (doseq [strat ["random" "optimal"]] (doTrials trials np strat))))))

(defn -main []
  (let [rt (Runtime/getRuntime)
    start-mem (- (.totalMemory rt) (.freeMemory rt))
    start (System/nanoTime)]
      (main)
      (System/gc)
      (let [end (System/nanoTime)
        end-mem (- (.totalMemory rt) (.freeMemory rt))
        duration-us (quot (- end start) 1000)
        memory-bytes (Math/abs ^long (- end-mem start-mem))]
        (println (str "{\n  \"duration_us\": " duration-us ",\n  \"memory_bytes\": " memory-bytes ",\n  \"name\": \"main\"\n}"))
      )
    ))

(-main)
